apps/reaper/lib/reaper/data_extract/extract_step.ex

```diff
@@ -0,0 +1,83 @@
+defmodule Reaper.DataExtract.ExtractStep do
+  @moduledoc """
+  This module processes extract steps as defined in a dataset definition.  After
+  iterating through the steps, accumilating any destination values in the assigns block
+  it is assumed the final step will be http (at this time) which returns a data stream
+  """
+  require Logger
+  alias Reaper.DataSlurper
+  alias Reaper.UrlBuilder
+  alias Reaper.Decoder
+
+  def execute_extract_steps(dataset, steps) do
+    Enum.reduce(steps, %{}, fn step, acc ->
+      execute_extract_step(dataset, step, acc)
+    end)
+  end
+
+  defp execute_extract_step(dataset, step, assigns_accumulator) do
+    step = Map.put(step, :assigns, Map.merge(step.assigns, assigns_accumulator))
+    process_extract_step(dataset, step)
+  rescue
+    error ->
+      Logger.error(Exception.format(:error, error, __STACKTRACE__))
+      reraise "Unable to process #{step.type} step for dataset #{dataset.id}.", __STACKTRACE__
+  end
+
+  defp process_extract_step(dataset, %{type: "http"} = step) do
+    headers = UrlBuilder.safe_evaluate_parameters(step.context.headers, step.assigns)
+
+    body =
+      UrlBuilder.safe_evaluate_parameters(step.context.body, step.assigns)
+      |> Enum.into(%{})
+      |> Jason.encode!()
+
+    UrlBuilder.decode_http_extract_step(step)
+    |> DataSlurper.slurp(dataset.id, headers, step.context.protocol, step.context.action, body)
+    |> Decoder.decode(dataset)
+    |> Stream.with_index()
+  end
+
+  defp process_extract_step(_dataset, %{type: "date"} = step) do
+    date =
+      case step.context.deltaTimeUnit do
+        nil ->
+          Timex.now()
+
+        _ ->
+          unit = String.to_atom(step.context.deltaTimeUnit)
+          Timex.shift(Timex.now(), [{unit, step.context.deltaTimeValue}])
+      end
+
+    formatted_date = Timex.format!(date, step.context.format)
+    Map.put(step.assigns, step.context.destination |> String.to_atom(), formatted_date)
```

Comment:
Some of these read a bit weird, but it's not a huge deal

Reply:
If anything, it's possible some of these could be dried up under a more expressive function name